sycl/doc/OptionalDeviceFeatures.md

```diff
@@ -0,0 +1,605 @@
+exception if it does not.
+
+When a kernel is submitted to a device, the runtime finds the device image that
+contains the kernel and also finds the kernel's entry in the
+"SYCL/requirements" property set.  This entry tells the set of requirements for
+the kernel.  If the target device does not support all of these requirements,
+then the runtime raises `errc::kernel_not_supported`.  This check can be done
+before the device image is JIT compiled, so the exception can be thrown
+synchronously.
+
+If the kernel imports device function symbols from a shared library as defined
+in [Device Code Dynamic Linking][7], the runtime first identifies all the
+device images that define these exported device functions.  Before attempting
+to link them together, the runtime finds the entries for the exported device
+functions in their "SYCL/requirements" property sets and checks that the device
+supports all these requirements.  If it does not, the runtime throws
+`errc::kernel_not_supported`.
+
+Whenever the runtime submits a SPIR-V image to the backend for online
+compilation, it must do the following additional steps:
+
+* Compute the union of all requirements from all entries in the image's
+  "SYCL/requirements" property set.
+
+* Query the target device to see whether it supports each of these
+  requirements, yielding either `True` or `False` for each one.
+
+* Set the value of each associated specialization constant when JIT compiling
+  the device image for this target device.
+
+Note that the runtime's cache of compiled device images does not need any
+special modification because the cache already needs to know the values of all
+the specialization constants that were used to compile the device image.  We
+just need to make sure the cache is also aware of the specialization constants
+which correspond to the requirements from the "SYCL/requirements" property set.
+
+#### Modifications to the GEN compiler
+
+The GEN compiler, of course, needs to be modified to implement the new
+`ConditionalINTEL` SPIR-V decoration.  It must discard any function with this
+decoration (unless the corresponding specialization constant is `True`), and it
+must not raise any sort of error due to compilation of these discarded
+functions.
+
+### AOT case
+
+The AOT case uses exactly the same solution as the JIT case described above,
+but there is one extra steps.  For the AOT case, the post-link tool must set
+the values of the specialization constants that correspond to the requirements
+for the kernel or exported device function, using the device named in the
+"-fsycl-targets" command line option.  After doing this, the post-link tool
+calls the AOT compiler to generate native code from SPIR-V as it normally does.
+If more than one target device is specified, the post-link tool sets the
+specialization constants separately for each device before generating native
+code for that device.
+
+Note that the native device image may not contain all kernels or all exported
```

Comment:
This is an important note which should be also somehow highlighted in `sycl-post-link` section, because it basically means do requirements collection before device code split and use gathered list of requirements for each resulting device image. I definitely missed this details while was reading about `sycl-post-link` changes.

Reply:
I think you may misunderstand what I'm saying, but I'm not sure.  I was thinking that `sycl-post-link` will gather the requirements **for each kernel** before device code split, but it doesn't really matter since the information is per-kernel.  Of course, `sycl-post-link` will need to construct the contents of the "SYCL/requirements" property set after device code split, since the property set is per-device image, and we don't know the kernels in each device image until after device code split.  I'll try to make this clearer in the document.

The thing I'm pointing out here is specific to the AOT case, where we run the SPIRV->native compiler at AOT time.  When this happens, the native device image may not contain all the kernels that were in the SPIR-V image because some of them may have been discarded due to the `ConditionalINTEL` SPIR-V decoration.

Reply:
I think this issue is resolved now, but please feel free to commend (and unresolve this thread) if you still have concerns.